File: app/src/main/java/com/github/android/lvrn/lvrnproject/persistent/repository/impl/BasicRepositoryImpl.java
package com.github.android.lvrn.lvrnproject.persistent.repository.impl;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.SQLException;
import android.database.sqlite.SQLiteDatabase;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.github.android.lvrn.lvrnproject.persistent.database.DatabaseManager;
import com.github.android.lvrn.lvrnproject.persistent.entity.Entity;
import com.github.android.lvrn.lvrnproject.persistent.repository.BasicRepository;
import com.google.common.base.Optional;
import com.orhanobut.logger.Logger;

import java.util.ArrayList;
import java.util.List;

import static com.github.android.lvrn.lvrnproject.persistent.database.LavernaContract.LavernaBaseTable.COLUMN_ID;

/**
 * @author Vadim Boitsov <vadimboitsov1@gmail.com>
 */

public abstract class BasicRepositoryImpl<T extends Entity>  implements BasicRepository<T> {
    private static final String TAG = "BasicRepoImpl";

    /**
     * A name of a table represented by the repository.
     */
    final String mTableName;

    @Nullable
    protected SQLiteDatabase mDatabase;

    public BasicRepositoryImpl(@NonNull String mTableName) {
        this.mTableName = mTableName;
    }

    @Override
    public boolean add(@NonNull T entity) {
        boolean result = false;
        mDatabase.beginTransaction();
        try {
            result = mDatabase.insert(mTableName, null, toContentValues(entity)) != -1;
            mDatabase.setTransactionSuccessful();
        } catch (SQLException e) {
            Logger.e(e, "Error while doing transaction.");
        } finally {
            mDatabase.endTransaction();
        }
        Logger.d("Table name: $s\nOperation: add\nEntity: $s\nResult: $s", mTableName, entity, result);
        return result;
    }

    @Override
    public boolean remove(@NonNull String id) {
        boolean result = false;
        mDatabase.beginTransaction();
        try {
            result = mDatabase.delete(mTableName, COLUMN_ID + "= '" + id + "'", null) != 0;
            mDatabase.setTransactionSuccessful();
        } catch (SQLException e) {
            Logger.e(e, "Error while doing transaction.");
        } finally {
            mDatabase.endTransaction();
        }
        Logger.d("Table name: $s\nOperation: remove\nId: $d\nResult: $s", mTableName, id, result);
        return result;
    }

    @NonNull
    @Override
    public Optional<T> getById(@NonNull String id) {
        Cursor cursor = mDatabase.rawQuery(
                "SELECT * FROM " + mTableName
                        + " WHERE " + COLUMN_ID + " = '" + id + "'",
                new String[]{});
        Logger.d("Table name: $s\nOperation: getById\nId: $d\nCursor: $s", mTableName, id, (cursor != null));
        if (cursor != null && cursor.moveToFirst()) {
            return Optional.of(toEntity(cursor));
        }
        return Optional.absent();
    }

    @Override
    public boolean openDatabaseConnection() {
        if (mDatabase != null) {
            Logger.w("Connection is already opened");
            return false;
        }
        mDatabase = DatabaseManager.getInstance().openConnection();
        Logger.i("Connection is opened");
        return true;
    }

    @Override
    public boolean closeDatabaseConnection() {
        if (mDatabase == null) {
            Logger.w("Connection is already closed");
            return false;
        }
        DatabaseManager.getInstance().closeConnection();
        mDatabase = null;
        Logger.i("Connection is closed");
        return true;
    }

    /**
     * A method which retrieves objects from the database by a raw SQL query.
     * @param query a {@code String} object of a raw SQL query
     * @return a lis of objects.
     */
    @NonNull
    protected List<T> getByRawQuery(String query) {
        Cursor cursor = mDatabase.rawQuery(query, new String[]{});
        Logger.d("Raw query: $s\nCursor: $s", query, (cursor != null));
        List<T> entities = new ArrayList<>();
        if (cursor != null && cursor.moveToFirst()) {
            do {
                entities.add(toEntity(cursor));
            } while (cursor.moveToNext());
            cursor.close();
        }
        return entities;
    }

    /**
     * A method which converts a received entity into a ContentValues object.
     * @param entity a {@code ProfileDependedEntity} extended object to convert.
     * @return a {@code ContentValues} object.
     */
    protected abstract ContentValues toContentValues(T entity);

    /**
     * A method which parse a received cursor result of a query into an entity.
     * @param cursor a {@code Cursor} with a data to convert into entity.
     * @return a {@code ProfileDependedEntity} object.
     */
    protected abstract T toEntity(Cursor cursor);
}

File: app/src/main/java/com/github/android/lvrn/lvrnproject/persistent/repository/impl/BasicRepositoryImpl.kt
package com.github.android.lvrn.lvrnproject.persistent.repository.impl

import android.content.ContentValues
import android.database.Cursor
import android.database.SQLException
import android.database.sqlite.SQLiteDatabase
import com.github.android.lvrn.lvrnproject.persistent.database.DatabaseManager.Companion.getInstance
import com.github.android.lvrn.lvrnproject.persistent.database.LavernaContract.LavernaBaseTable
import com.github.android.lvrn.lvrnproject.persistent.repository.BasicRepository
import com.github.valhallalabs.laverna.persistent.entity.base.Entity
import com.orhanobut.logger.Logger
import java.util.*

abstract class BasicRepositoryImpl<T : Entity>(val tableName: String) : BasicRepository<T> {

    /**
     * A name of a table represented by the repository.
     */
    protected var mDatabase: SQLiteDatabase? = null

    override fun add(entity: T): Boolean {
        var result = false
        mDatabase!!.beginTransaction()
        try {
            result = mDatabase!!.insertOrThrow(tableName, null, toContentValues(entity)) != -1L
            mDatabase!!.setTransactionSuccessful()
        } catch (e: SQLException) {
            Logger.e(e, "Error while doing transaction.")
        } finally {
            mDatabase!!.endTransaction()
        }
        Logger.d("Table name: %s\nOperation: add\nEntity: %s\nResult: %s", tableName, entity, result)
        return result
    }

    override fun remove(id: String): Boolean {
        var result = false
        mDatabase!!.beginTransaction()
        try {
            result = mDatabase!!.delete(tableName, LavernaBaseTable.COLUMN_ID + "= '" + id + "'", null) != 0
            mDatabase!!.setTransactionSuccessful()
        } catch (e: SQLException) {
            Logger.e(e, "Error while doing transaction.")
        } finally {
            mDatabase!!.endTransaction()
        }
        Logger.d("Table name: %s\nOperation: remove\nId: %s\nResult: %s", tableName, id, result)
        return result
    }

    override fun getById(id: String): Optional<T> {
        val cursor = mDatabase!!.rawQuery(
                "SELECT * FROM " + tableName
                        + " WHERE " + LavernaBaseTable.COLUMN_ID + " = '" + id + "'", arrayOf())
        Logger.d("Table name: %s\nOperation: getById\nId: %s\nCursor: %s", tableName, id, cursor != null)
        return if (cursor != null && cursor.moveToFirst()) {
            Optional.of(toEntity(cursor))
        } else Optional.empty()
    }

    override fun openDatabaseConnection(): Boolean {
        if (mDatabase != null) {
            Logger.w("Connection is already opened")
            return false
        }
        mDatabase = getInstance()!!.openConnection()
        Logger.i("Connection is opened")
        return true
    }

    override fun closeDatabaseConnection(): Boolean {
        if (mDatabase == null) {
            Logger.w("Connection is already closed")
            return false
        }
        getInstance()!!.closeConnection()
        mDatabase = null
        Logger.i("Connection is closed")
        return true
    }

    override fun isConnectionOpened(): Boolean {
        return mDatabase != null
    }

    /**
     * A method which retrieves objects from the database by a raw SQL query.
     * @param query a `String` object of a raw SQL query
     * @return a list of objects.
     */
    protected open fun getByRawQuery(query: String?): MutableList<T> {
        val cursor = mDatabase!!.rawQuery(query, arrayOf())
        Logger.d("Raw query: %s\nCursor: %s", query, cursor != null)
        val entities: MutableList<T> = ArrayList()
        if (cursor != null && cursor.moveToFirst()) {
            do {
                entities.add(toEntity(cursor))
            } while (cursor.moveToNext())
            cursor.close()
        }
        return entities
    }

    /**
     * A method which converts a received entity into a ContentValues object.
     * @param entity a `ProfileDependedEntity` extended object to convert.
     * @return a `ContentValues` object.
     */
    protected abstract fun toContentValues(entity: T): ContentValues

    /**
     * A method which parse a received cursor result of a query into an entity.
     * @param cursor a `Cursor` with a data to convert into entity.
     * @return a `ProfileDependedEntity` object.
     */
    protected abstract fun toEntity(cursor: Cursor): T
}